### PR TITLE
Add 'mvt' field type format to geo fields

### PR DESCRIPTION
**Pull request by iverase**
In #73872 we have introduced a new end point that produces vector tiles. One of the actions on this new end point is to transform geo data expressed in WKT or geojson into vector tiles features. It currently works by sung the fields API to get the geo data in geojson format and then transform it to a vector tile feature while serialising the result of the query into a vector tile.

This is currently not very efficient; We need to parse twice the shape, the first time the fields API reads the shape from source and serialise it into geojson. Then we parse it again when building the vector tile and transform it to a vector tile feature. Hence we are parsing twice the same geometry.

In this PR we propose to extend the Fields API formats for geo data in order to produce vector tiles features directly, instead of using geojson as intermediate format. It is easy to show that this bring much better performance overall and in particular is very beneficial because the compression of complex geometries when using mvt features (up to 400+ times smaller).

The main changes in this PR are:

* Move The mvt factory for points to server so it can be used by geo_point fields. This factory has no dependencies with external libraries.

* Add a dependency between the spatial module and the vector tile module so we can access the mvt factory for geometries for geo_shape fields.

* mvt needs some extra information in order to encode the geometry. That is the current tile and the extent (number of pixels) of the tile. The fields API does not allow to add this information easily so we take the following approach. In order to to support extra information, we allow to declare mvt formats with the following formats: mvt(z/x/y@extent) or mvt(z/x/y) (in the last one, extent is considered the default one: 4096). The logic to parse the new format is in GeoFormatterFactory.

Some results of this PR.

Using a data set that contains few big polygons with thousand of points each, we have run the following query for geojson and mvt(0/0/0) formats.

```
curl -X GET "localhost:9200/example/_search"  -H 'Content-Type: application/json' -d '
{
  "size" : 10000,
  "_source" : false,
  "fields" : [{"field" :"location", "format" : "{format}"}],
  "runtime_mappings" : {
      "width" : {
         "script" : "emit(doc[\"location\"].getBoundingBox().bottomRight().getLon() - doc[\"location\"].getBoundingBox().topLeft().getLon())",
         "type" : "double"
      }
   },
   "sort" : [
     {
       "width" : {"order" : "desc"}
     }
   ]
}'
```

When running with geojson, the resulting json output has a size of 438.5Mb. In addition I had to increase the heap size of the one node cluster to 4GB in order not to OOM.
When running with mvt(0/0/0), the resulting json output has a size of 1.1MB. It was run with a heap of 1GB and there was no signs of GC pressure.

When running similar queries through the _mvt endpoint, we can observe that with this change the request is twice as fast as the current one (5 min vs 2.5 min). In addition it takes much less memory to complete.

relates to https://github.com/elastic/elasticsearch/pull/74476

### CODE REVIEW

File: x-pack/plugin/spatial/build.gradle [hunk 1]
```diff
@@ -15,6 +15,7 @@ dependencies {
   testImplementation(testArtifact(project(xpackModule('core'))))
   yamlRestTestImplementation(testArtifact(project(xpackModule('core'))))
   api project(path: ':modules:geo')
+  api project(path: xpackModule('vector-tile'))
```
Comment (iverase) [inline x-pack/plugin/spatial/build.gradle:18]: Is this enough and ok?

Comment (colings86) [reply x-pack/plugin/spatial/build.gradle:18]: If you are asking from the build POV I think @pugnascotia or @breskeby might be able to assist here

**Review by colings86 — COMMENTED** [review]
From the point of view of the general approach of getting access to the Geo formatter I think this PR is good. I have not reviewed the change holistically though, just from that aspect

File: x-pack/plugin/spatial/build.gradle (hunk 1 above)
Comment (iverase) [reply x-pack/plugin/spatial/build.gradle:18]: Hey @elastic/es-delivery, do you have any feedback about the change above? We need to access a factory that lives in the vector-tile module from the spatial module.

File: server/src/main/java/org/elasticsearch/common/geo/GeoFormatterFactory.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
```
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/common/geo/GeoFormatterFactory.java:14]: nit: Sorry for the drive-by-comment, I was just scanning this PR coming from #74476 which this seems to superseed and was just wondering about the two copyright notices here and if one of them was maybe added by accident.

File: server/src/main/java/org/elasticsearch/common/geo/SimpleFeatureFactory.java [hunk 3]
```diff
@@ -1,25 +1,32 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
```
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/common/geo/SimpleFeatureFactory.java:7]: nit: same here, see the two copyright notices here

File: server/src/main/java/org/elasticsearch/common/geo/SphericalMercatorUtils.java [hunk 4]
```diff
@@ -1,18 +1,26 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
```
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/common/geo/SphericalMercatorUtils.java:7]: nit: same here, see the two copyright notices here. There are a few more cases in this PR so I'm leaving it at that in case this is done on purpose. Sorry for the noise in this case.

**Review by cbuescher — COMMENTED** [review]
I was just scanning this PR coming from #74476 which this seems to superseed and was just wondering about the two copyright notices here and if one of them was maybe added by accident. Sorry for the noise if this was done on purpose, I'd be interested in the policy we have around it in that case.

File: server/src/main/java/org/elasticsearch/common/geo/GeoFormatterFactory.java (hunk 2 above)
Comment (iverase) [reply server/src/main/java/org/elasticsearch/common/geo/GeoFormatterFactory.java]: This is a leftover

File: x-pack/plugin/spatial/build.gradle (hunk 1 above)
Comment (pugnascotia) [reply x-pack/plugin/spatial/build.gradle:18]: I defer to @breskeby here. We mostly have cross-plugin deps for test purposes, as far as I can tell, but this is probably OK. We could move it to `core`, but I don't if that would be worth it?

Comment (iverase) [reply x-pack/plugin/spatial/build.gradle:18]: I guess it would make more sense to move vector-tiles module to spatial as it is currently just adding the vector tile end point. It was originally included in the spatial module but we decided to move it its own module after discussing with Ryan. Maybe it wasn't the right decision?

Comment (breskeby) [reply x-pack/plugin/spatial/build.gradle:18]: from the build perspective this is fine. I don't have much background in how comfortable we are with cross plugin dependencies though.

Comment (mark-vieira) [reply x-pack/plugin/spatial/build.gradle:18]: I think this is all that's needed. These (despite the folder naming) are modules, not plugins, so there shouldn't be anything else needed here since that module is always guaranteed to be installed. I don't have strong opinions on how these two modules are otherwise structured from a modularity standpoint. It might very well make sense to combine them as @iverase suggests.

**Comment by iverase** [comment]
In this iteration we changed how the modules depends between them, so repeating the main changes of this PR:

- Move The mvt factory for points to server so it can be used by geo_point fields. This factory has no dependencies with external libraries.

- Add a dependency between the spatial module and the vector tile module so we can access the mvt factory for geometries for geo_shape fields. In particular now the vector tiles module extends the spatial module and using the extensibility plugin it register the vector tile engine in the spatial module (we only allow one vector tile engine to be register). We have change the typeParser for geo_shape so it can take the provided vector tile engine. 

- mvt needs some extra information in order to encode the geometry. That is the current tile and the extent (number of pixels) of the tile. The fields API does not allow to add this information easily so we take the following approach. In order to to support extra information, we allow to declare mvt formats with the following formats: mvt(z/x/y@extent) or mvt(z/x/y) (in the last one, extent is considered the default one: 4096). The logic to parse the new format is in GeoFormatterFactory.

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/VectorTileExtension.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+import org.elasticsearch.common.geo.GeoFormatterFactory;
+import org.elasticsearch.geometry.Geometry;
+
+
+public interface VectorTileExtension {
+    /**
+     * Get the vector tile engine. This is called when user ask for the MVT format on the field API.
+     * We are only expecting one instance of a vector tile engine coming from the vector tile module.
+     */
+    GeoFormatterFactory.VectorTileEngine<Geometry> getVectorTileEngine();
```
Comment (imotov) [inline x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/VectorTileExtension.java:19]: We are leaking way too much here. I think the right abstraction would be to just extend on the GeometryFromatter layer and hide the rest inside the tile class, but I think we can address it as another iteration.

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/SpatialPlugin.java [hunk 6]
```diff
@@ -205,4 +210,10 @@ private static void registerCardinalityAggregator(ValuesSourceRegistry.Builder b
             return realParser.parse(parser, name);
         };
     }
+
+    @Override
+    public void loadExtensions(ExtensionLoader loader) {
+        // we only expect one vector tile extension that comes from the vector tile module.
+        loader.loadExtensions(VectorTileExtension.class).forEach(vectorTileExtension::set);
```
Comment (imotov) [inline x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/SpatialPlugin.java:217]: That essentially moves the dependency from the compile time into runtime. It is not explicit, but the whole thing will break if spatial module is present but vector-tile module is not or if we have another alternative implementation for that. We should probably address this as a follow up as well.

**Review by imotov — APPROVED** [review]
LGTM in general. There are a couple of follow ups that I think we should do to make it less brittle and remove some vector-tile specific abstractions from the formatter.

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/SpatialPlugin.java [hunk 7]
```diff
@@ -84,7 +89,7 @@ protected XPackLicenseState getLicenseState() {
         Map<String, Mapper.TypeParser> mappers = new HashMap<>(super.getMappers());
         mappers.put(ShapeFieldMapper.CONTENT_TYPE, ShapeFieldMapper.PARSER);
         mappers.put(PointFieldMapper.CONTENT_TYPE, PointFieldMapper.PARSER);
-        mappers.put(GeoShapeWithDocValuesFieldMapper.CONTENT_TYPE, GeoShapeWithDocValuesFieldMapper.PARSER);
+        mappers.put(GeoShapeWithDocValuesFieldMapper.CONTENT_TYPE, new GeoShapeWithDocValuesFieldMapper.TypeParser(vectorTileExtension));
```
[the remark this answers is not in the document]
Comment (iverase) [reply x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/SpatialPlugin.java]: ok, I introduced a Boolean variable to make sure the extensions are loaded before calling the mappers. I did have the same feeling that SetOnce should not be in the mappers.

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/VectorTileExtension.java (hunk 5 above)
Comment (iverase) [reply x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/VectorTileExtension.java:19]: Not sure what you mean here but let's leave it for next iteration

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/SpatialPlugin.java (hunk 6 above)
Comment (iverase) [reply x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/SpatialPlugin.java:222]: My hope here is that if the vector-tile module is not present everything should just work. You just end up with an error if you request geometries in mvt format.

Comment (imotov) [reply x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/SpatialPlugin.java:222]: Yes, I stand corrected, it looks like it throws "vector tile format is not supported" exception in this case, which is hardcoded in the spatial module. But I think there is still too much knowledge to vector tile implementation inside spatial, let me try to take a shot at confine more of it into the vector tile module.

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/SpatialPlugin.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (iverase) [reply x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/SpatialPlugin.java]: I need to remove the assertion as some test has custom code to load mappers. I think integration test should still caught the case when extensions are not loaded before mappers so we are good.

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/SpatialPlugin.java (hunk 6 above)
Comment (iverase) [reply x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/SpatialPlugin.java:217]: That would be awesome. I ma going to push it as it is and we can iterate in a follow up.
